lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -25,78 +25,240 @@
-
-        # we could not use opensearch for locking, we use the peer rel data bag
-        self._charm.peers_data.put(Scope.UNIT, OpenSearchOpsLock.PEER_DATA_LOCK_FLAG, True)
-
-    def release(self):
-        """Method for Releasing the "ops" lock."""
-        host = self._charm.unit_ip if self._opensearch.is_node_up() else None
-
-        # can use opensearch to remove lock
-        if host is not None or self._charm.alt_hosts:
+                if (
+                    e.response_code == 400
+                    and e.response_body.get("error", {}).get("type")
+                    == "resource_already_exists_exception"
+                ):
+                    # Index already created
+                    pass
+                else:
+                    logger.exception("Error creating OpenSearch lock index")
+                    return False
+            # Attempt to create document id 0
             try:
                 self._opensearch.request(
-                    "DELETE",
-                    endpoint=f"/{OpenSearchOpsLock.LOCK_INDEX}",
+                    "PUT",
+                    endpoint=f"/{self._OPENSEARCH_INDEX}/_create/0?refresh=true",
                     host=host,
-                    alt_hosts=self._charm.alt_hosts,
+                    alt_hosts=alt_hosts,
                     retries=3,
+                    payload={"unit-name": self._charm.unit.name},
                 )
             except OpenSearchHttpError as e:
-                # ignore 404, it means the index is not found and this just means that
-                # the cleanup happened before but event got deferred because of another error
-                if e.response_code != 404:
-                    raise
-
-        self._charm.peers_data.delete(Scope.UNIT, OpenSearchOpsLock.PEER_DATA_LOCK_FLAG)
+                if e.response_code == 409 and "document already exists" in e.response_body.get(
+                    "error", {}
+                ).get("reason", ""):
+                    # Document already created
+                    if (unit := self._unit_with_lock(host)) != self._charm.unit.name:
+                        # Another unit has lock
+                        # (Or document deleted after last request & before request in
+                        # `self._lock_acquired()`)
+                        logger.debug(
+                            f"[Node lock] Not acquired. Unit with opensearch lock: {unit}"
+                        )
+                        return False
+                else:
+                    logger.exception("Error creating OpenSearch lock document")
+                    return False
+            # Lock acquired
+            # Release peer databag lock, if any
+            logger.debug("[Node lock] Acquired via opensearch")
+            self._peer.release()
+            logger.debug("[Node lock] Released redundant peer lock (if held)")
+            return True
+        else:
```

Comment:
nit: could we please invert the if/else so that we reduce a level of nestedness?
```
if not host and not alt_hosts:
    return self._peer.acquired

# rest of the code
```